zerver/lib/push_notifications.py

```diff
@@ -454,120 +460,88 @@ def send_android_push_notification(
         )
         return 0
 
     if remote:
         logger.info(
-            "GCM: Sending notification for remote user %s:%s to %d devices",
+            "FCM: Sending notification for remote user %s:%s to %d devices",
             remote.uuid,
             user_identity,
             len(devices),
         )
     else:
         logger.info(
-            "GCM: Sending notification for local user %s to %d devices", user_identity, len(devices)
+            "FCM: Sending notification for local user %s to %d devices", user_identity, len(devices)
         )
-    reg_ids = [device.token for device in devices]
-    priority = parse_gcm_options(options, data)
-    try:
-        # See https://firebase.google.com/docs/cloud-messaging/http-server-ref .
-        # Two kwargs `retries` and `session` get eaten by `json_request`;
-        # the rest pass through to the GCM server.
-        #
-        # One initial request plus 2 retries, with 5-second timeouts,
-        # and expected 1 + 2 seconds (the gcm module jitters its
-        # backoff by ±50%, so worst case * 1.5) between them, totals
-        # 18s expected, up to 19.5s worst case.
-        res = gcm_client.json_request(
-            registration_ids=reg_ids,
-            priority=priority,
-            data=data,
-            retries=2,
-            session=FCMSession(),
+
+    token_list = [device.token for device in devices]
+    priority = parse_fcm_options(options, data)
+
+    # The API requires all values to be strings. Our data dict is going to have
+    # things like an integer realm and user ids etc., so just convert everything
+    # like that.
+    data = {k: str(v) if not isinstance(v, str) else v for k, v in data.items()}
+    messages = [
+        messaging.Message(
+            data=data, token=token, android=messaging.AndroidConfig(priority=priority)
         )
-    except OSError:
-        logger.warning("Error while pushing to GCM", exc_info=True)
-        return 0
+        for token in token_list
+    ]
 
-    successfully_sent_count = 0
-    if res and "success" in res:
-        for reg_id, msg_id in res["success"].items():
-            logger.info("GCM: Sent %s as %s", reg_id, msg_id)
-        successfully_sent_count = len(res["success"].keys())
+    try:
+        batch_response = messaging.send_each(messages, app=fcm_app)
+    except exceptions.FirebaseError:
+        logger.warning("Error while pushing to FCM", exc_info=True)
+        return 0
 
     if remote:
         assert settings.ZILENCER_ENABLED
         DeviceTokenClass: Type[AbstractPushDeviceToken] = RemotePushDeviceToken
     else:
         DeviceTokenClass = PushDeviceToken
 
-    # res.canonical will contain results when there are duplicate registrations for the same
-    # device. The "canonical" registration is the latest registration made by the device.
-    # Ref: https://developer.android.com/google/gcm/adv.html#canonical
-    if "canonical" in res:
```

Comment:
The `canonical` case disappears, because as discussed, the concept doesn't seem to exist in the new API, and these cases don't seem to occur anymore even in the legacy API

Reply:
Maybe rip this out as a prep commit?